src/backend/distributed/utils/metadata_cache.c

```diff
@@ -911,6 +968,119 @@ InitializeDistTableCache(void)
 
 
 /*
+ * GetWorkerNodeHash is a wrapper around InitializeWorkerNodeCache(). It
+ * triggers InitializeWorkerNodeCache when the workerHash is NULL. Otherwise,
+ * it returns the hash.
+ */
+HTAB *
+GetWorkerNodeHash(void)
+{
+	if (!workerNodeHashValid)
+	{
+		InitializeWorkerNodeCache();
+
+		workerNodeHashValid = true;
+	}
+
+	return WorkerNodeHash;
+}
+
+
+/*
+ * Initialize the infrastructure for the worker node cache. The functions
+ * reads the worker nodes from the metadata table, adds them to the hash and
+ * finally registers an invalidation callaback.
+ */
+static void
+InitializeWorkerNodeCache(void)
+{
+	static bool invalidationRegistered = false;
+	HTAB *oldWorkerNodeHash = NULL;
+	HTAB *newWorkerNodeHash = NULL;
+	List *workerNodeList = NIL;
+	ListCell *workerNodeCell = NULL;
+	HASHCTL info;
+	int hashFlags = 0;
+	long maxTableSize = (long) MaxWorkerNodesTracked;
+
+	/* make sure we've initialized CacheMemoryContext */
+	if (CacheMemoryContext == NULL)
+	{
+		CreateCacheMemoryContext();
+	}
+
+	/*
+	 * Create the hash that holds the worker nodes. The key is the unique nodeid
+	 * field.
+	 */
+	memset(&info, 0, sizeof(info));
+	info.keysize = sizeof(uint32);
+	info.entrysize = sizeof(WorkerNode);
+	info.hcxt = CacheMemoryContext;
+	info.hash = tag_hash;
+	hashFlags = HASH_ELEM | HASH_FUNCTION | HASH_CONTEXT;
+
+	newWorkerNodeHash = hash_create("Worker Node Hash",
+									maxTableSize,
+									&info, hashFlags);
+
+	/* read the list from the pg_dist_node */
+	workerNodeList = ReadWorkerNodes();
+
+	/* iterate over the worker node list */
+	foreach(workerNodeCell, workerNodeList)
+	{
+		WorkerNode *workerNode = NULL;
+		WorkerNode *currentNode = lfirst(workerNodeCell);
+		void *hashKey = NULL;
+		bool handleFound = false;
+
+		/*
+		 * Search for the worker node in the hash, and then insert the
+		 * values. When searching, we make the hashKey the unique nodeid.
+		 */
+		hashKey = (void *) &currentNode->nodeId;
+		workerNode = (WorkerNode *) hash_search(newWorkerNodeHash, hashKey,
+												HASH_ENTER, &handleFound);
+
+		/* fill the newly allocated workerNode in the cache */
+		strlcpy(workerNode->workerName, currentNode->workerName, WORKER_LENGTH);
+		workerNode->workerPort = currentNode->workerPort;
+		workerNode->groupId = currentNode->groupId;
+
+		if (handleFound)
+		{
+			ereport(WARNING, (errmsg("multiple lines for worker node: \"%s:%u\"",
+									 workerNode->workerName,
+									 workerNode->workerPort)));
+		}
+
+		workerNode->workerPort = currentNode->workerPort;
+
+		/* we do not need the currentNode anymore */
+		pfree(currentNode);
+	}
+
+	/* assign the newly created hash */
+	oldWorkerNodeHash = WorkerNodeHash;
+	WorkerNodeHash = newWorkerNodeHash;
+
+	/* now, safe to destroy the old hash */
+	hash_destroy(oldWorkerNodeHash);
```

Comment:
We don't really need the distinction between newWorkerNodeHash and oldWorkerNodeHash.